File: codility/lessons/10_prime_and_composite_numbers/flags.py
# https://app.codility.com/programmers/lessons/10-prime_and_composite_numbers/flags/

def is_peak(A, index):
    return 0 < index < len(A)-1 and A[index-1] < A[index] > A[index+1]

def min_distance_peaks_count(peaks, min_distance):
    if not peaks:
        return 0

    peaks_count = 0

    index = 0
    while index < len(peaks):
        next_index = index + 1
        while next_index < len(peaks) and peaks[index] + min_distance > peaks[next_index]:
            next_index += 1
        if next_index < len(peaks):
            peaks_count += 1
        index = next_index

    return peaks_count


def solution(A):
    peaks_bitmap = (is_peak(A, index) for index, _ in enumerate(A))
    peaks = [peak for peak, _ in filter(lambda x: x[1], enumerate(peaks_bitmap))]

    # Intead of looking for maximum k iteratievly do a bisection.
    start_k, end_k = 1, len(peaks) + 1
    while start_k < end_k:
        mid_k = (start_k + end_k) // 2
        if mid_k <= min_distance_peaks_count(peaks, mid_k):
            start_k = mid_k + 1
        else:
            end_k = mid_k

    k = start_k - 1
    return k

File: codility/lessons/10_prime_and_composite_numbers/flags.cpp
#include <algorithm>
#include <vector>
using namespace std;

inline bool is_peak(vector<int> &A, size_t i) {
    return A[i - 1] < A[i] && A[i] > A[i + 1];
}

int solution(vector<int> &A) {
    size_t flags_begin = 0;
    size_t flags_end = A.size();

    while (flags_begin < flags_end) {
        size_t flags_taken = flags_begin + (flags_end - flags_begin)/2;
        size_t min_distance = flags_taken;
        size_t flags_used = 0;

        for (size_t i = 1; i < A.size() - 1 && flags_used < flags_taken; ) {
            if (is_peak(A, i)) {
                flags_used++;
                i += min_distance;
            } else {
                i++;
            }
        }
        if (flags_used < flags_taken) {
            flags_end = flags_taken;
        } else {
            flags_begin = flags_taken + 1;
        }

    }

    return flags_begin - 1;
}
